spring-data-mongodb/src/main/java/org/springframework/data/mongodb/MongoExpression.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.mongodb;
+
+import java.util.function.Function;
+
+/**
+ * Wrapper object for MongoDB expressions like {@code $toUpper : $name} that manifest as {@link org.bson.Document} when
+ * passed on to the driver.
+ * <p />
+ * A set of predefined {@link MongoExpression expressions}, including a
+ * {@link org.springframework.data.mongodb.core.aggregation.AggregationSpELExpression SpEL based variant} for method
+ * like expressions (eg. {@code toUpper(name)}) are available via the
+ * {@link org.springframework.data.mongodb.core.aggregation Aggregation API}.
+ *
+ * @author Christoph Strobl
+ * @since 3.2
+ * @see org.springframework.data.mongodb.core.aggregation.ArithmeticOperators
+ * @see org.springframework.data.mongodb.core.aggregation.ArrayOperators
+ * @see org.springframework.data.mongodb.core.aggregation.ComparisonOperators
+ * @see org.springframework.data.mongodb.core.aggregation.ConditionalOperators
+ * @see org.springframework.data.mongodb.core.aggregation.ConvertOperators
+ * @see org.springframework.data.mongodb.core.aggregation.DateOperators
+ * @see org.springframework.data.mongodb.core.aggregation.ObjectOperators
+ * @see org.springframework.data.mongodb.core.aggregation.SetOperators
+ * @see org.springframework.data.mongodb.core.aggregation.StringOperators @
+ */
+@FunctionalInterface
+public interface MongoExpression {
+
+	/**
+	 * Obtain the native {@link org.bson.Document} representation.
+	 *
+	 * @return never {@literal null}.
+	 */
+	org.bson.Document toDocument();
+
+	/**
+	 * Convert this instance to another expression applying the given conversion {@link Function}.
+	 *
+	 * @param function must not be {@literal null}.
+	 * @param <T>
+	 * @return never {@literal null}.
+	 */
+	default <T extends MongoExpression> T as(Function<MongoExpression, T> function) {
+		return function.apply(this);
+	}
+
+	/**
+	 * Create a new {@link MongoExpression} from plain String (eg. {@code $toUpper : $name}). <br />
+	 * The given source value will be wrapped with <code>{ }</code> to match an actual MongoDB {@link org.bson.Document}
+	 * if necessary.
+	 *
+	 * @param json must not be {@literal null}.
+	 * @return new instance of {@link MongoExpression}.
+	 */
+	static MongoExpression expressionFromString(String json) {
```

Comment:
How about `parse`?

Reply:
Two things: 
- static import clashes with eg. `org.bson.Document.parse` 
- the expression must not be evaluated immediately which is something I'd expect from `parse`.